Initial OpenACC port of atm_bdy_adjust_dynamics_relaxzone_tend

abishekg7 opened the pull request:
Initial OpenACC port of atm_bdy_adjust_dynamics_relaxzone_tend

gdicker1 commented on src/core_atmosphere/dynamics/mpas_atm_time_integration.F:
```diff
@@ -6579,14 +6619,18 @@ subroutine atm_bdy_adjust_dynamics_relaxzone_tend( config, tend, state, diag, me
+      !$acc parallel default(present)
+      !$acc loop gang worker private(cell1, cell2, vertex1, vertex2, r_dc, r_dv, & 
+      !$acc    iCell, iVertex, invArea, iEdge_div, iEdge_vort, edge_sign, &
+      !$acc    laplacian_filter_coef, divergence1, divergence2, vorticity1, vorticity2)
```
Here I'd only expect the arrays (`divergence1`, `divergence2`, `vorticity1`, and `vorticity2`) to be declared private. The rest should automatically be private at the worker level and passed on to their (vector) threads. Are you being explicit here, or do you need to also add the scalar-values to the private clause for correctness?

abishekg7 replied:
True. I briefly didn't remember that about scalars when I wasn't getting bit identical results. Will make these changes and try again.

gdicker1 commented on src/core_atmosphere/dynamics/mpas_atm_time_integration.F:
```diff
@@ -6579,14 +6619,18 @@ subroutine atm_bdy_adjust_dynamics_relaxzone_tend( config, tend, state, diag, me
          if ( (bdyMaskEdge(iEdge) > 1) .and. (bdyMaskEdge(iEdge) <= nRelaxZone) ) then ! relaxation zone
 
-            laplacian_filter_coef = dcEdge(iEdge)**2 * (real(bdyMaskEdge(iEdge)) - 1.)/   &
+            laplacian_filter_coef = dcEdge(iEdge)*dcEdge(iEdge)* (real(bdyMaskEdge(iEdge)) - 1.)/   &
```
Pure curiosity about this change: could you explain your motivation a little? Does this have any effect?

abishekg7 replied:
I was trying this out because I though it may be one of the obstacles to bit identical results - but probably not. Will remove this change. Although, I would also add that we could just be using the multiplication operator here instead of exponent. But that's for another time.

gdicker1 replied:
Agreed, I think this may be better as a multiplication than exponent but that goes against keeping to a "light touch" in this phase.

mgduda replied:
Compilers are reasonably good at spotting these opportunities for optimization. The `nvfortran` compiler turns this line:
```
pure function square(x)

    real, intent(in) :: x
    real :: square

    square = x**2

end function square
```
into this assembler:
```
	.text
	.globl	square_                         # -- Begin function square_
	.type	square_,@function
square_:                                # @square_
	vmovss	(%rdi), %xmm0                   # xmm0 = mem[0],zero,zero,zero
	vmulss	%xmm0, %xmm0, %xmm0
	retq
```

gdicker1 reviewed (changes requested):
Besides a more formal commit message, here's some things I've noticed so far.

jim-p-w commented on src/core_atmosphere/dynamics/mpas_atm_time_integration.F:
```diff
@@ -6529,37 +6550,55 @@ subroutine atm_bdy_adjust_dynamics_relaxzone_tend( config, tend, state, diag, me
             end do
          end if
       end do
+      !$acc end parallel
 
+      !$acc parallel default(present)
```
Can both of these loops be in the same parallel region and run simultaneously? They appear to be orthogonal to each other. That is, neither loop references the array(s) the other loop changes.

abishekg7 replied:
Fusing these two parallel regions does work, but I'm wondering if we should perhaps leave this for the optimization stage.

jim-p-w replied:
I am not knowledgable enough to weigh in on that. But if it succeeded and BFB tests passed I'd put it in now.
I'm fine with it either way, I'll leave it to you, Dylan, and Michael to decide.

mgduda replied:
I would support the fusion of these two parallel regions in this PR, especially since doing so would lead to fewer lines in the diff.

jim-p-w commented on src/core_atmosphere/dynamics/mpas_atm_time_integration.F:
```diff
@@ -6529,37 +6550,55 @@ subroutine atm_bdy_adjust_dynamics_relaxzone_tend( config, tend, state, diag, me
 
             laplacian_filter_coef = (real(bdyMaskCell(iCell)) - 1.)/real(nRelaxZone)/(10.*dt*meshScalingRegionalCell(iCell))
             !
+            !$acc loop seq
```
Given that local scalars are automatically privatized, it seems this loop doesn't have to execute sequentially.

abishekg7 replied:
So there's a summation going on in the innermost loop for `tend_rt` and `tend_rho`, and ignoring the order of summation, there would still be an incorrect summation if `do i=1,nEdgesOnCell(iCell)` was parallelized. 

Now we could re-write the loop to use the `acc reduction(+: tend_rt, tend_rho)` clause, but we decided to keep the code changes minimal in this round of porting.

jim-p-w replied:
I missed that! Thanks for double checking.

jim-p-w commented on src/core_atmosphere/dynamics/mpas_atm_time_integration.F:
```diff
@@ -6579,14 +6619,18 @@ subroutine atm_bdy_adjust_dynamics_relaxzone_tend( config, tend, state, diag, me
       !  Third (and last), the horizontal filter for ru
-
+      !$acc parallel default(present)
+      !$acc loop gang worker private(cell1, cell2, vertex1, vertex2, r_dc, r_dv, & 
```
Local scalar vars are automatically privatized. Additionally, there's a performance hit when explicitly declaring them private, as it causes them to be placed in global memory. [See this nvidia forum item](https://forums.developer.nvidia.com/t/openacc-private-variables/133287)

abishekg7 replied:
Thanks, yeah I'll remove the scalars from private. And did not know this about global memory - thanks!

jim-p-w commented on src/core_atmosphere/dynamics/mpas_atm_time_integration.F:
```diff
@@ -6579,14 +6619,18 @@ subroutine atm_bdy_adjust_dynamics_relaxzone_tend( config, tend, state, diag, me
+      !$acc parallel default(present)
+      !$acc loop gang worker private(cell1, cell2, vertex1, vertex2, r_dc, r_dv, & 
+      !$acc    iCell, iVertex, invArea, iEdge_div, iEdge_vort, edge_sign, &
+      !$acc    laplacian_filter_coef, divergence1, divergence2, vorticity1, vorticity2)
```
Why does `laplacian_filter_coef` have to be private in this loop? It's not declared private in the other parallel loops in this subroutine.

abishekg7 commented on src/core_atmosphere/dynamics/mpas_atm_time_integration.F:
```diff
@@ -6529,37 +6550,55 @@ subroutine atm_bdy_adjust_dynamics_relaxzone_tend( config, tend, state, diag, me
+      call mpas_pool_get_config(config, 'config_relax_zone_divdamp_coef', divdamp_coef_ptr)
+
+      divdamp_coef = divdamp_coef_ptr
+      vertexDegree = vertexDegree_ptr
```
Will do.

abishekg7 replied:
Thanks! Done.

gdicker1 commented:
This looks good to me now. Just waiting for any changes from Michael's review and the final commit(s)

mgduda reviewed (approved):
Looks good to me.

@abishekg7 I'll merge this after you've reworked the commit history.

abishekg7 commented:
> Looks good to me.
> 
> @abishekg7 I'll merge this after you've reworked the commit history.

Thanks! I've rebased and cleaned up the commit history. And results are still bit identical with develop.
